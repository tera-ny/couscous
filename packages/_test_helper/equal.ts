import {
  assertEquals,
  equal,
} from "https://deno.land/std@0.132.0/testing/asserts.ts";
import { copy } from "https://deno.land/std@0.132.0/fs/mod.ts";
export const assertSnapshot = async (
  receivedPath: string,
  expectedPath: string
) => {
  const [received, expected] = await Promise.all([
    Deno.readTextFile(receivedPath),
    Deno.readTextFile(expectedPath),
  ]);
  if (Deno.args.includes("update_snapshot")) {
    if (equal(received, expected)) return;
    await copy(receivedPath, expectedPath, { overwrite: true });
  } else {
    assertEquals(received, expected);
  }
};
