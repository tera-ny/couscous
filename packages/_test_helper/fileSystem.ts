import {
  Project,
  SourceFile,
  ManipulationSettings,
  IndentationText,
} from "https://deno.land/x/ts_morph@14.0.0/mod.ts";

export const withDir = async (
  cb: (tempDir: string) => void | Promise<void>
) => {
  const tempDir = await Deno.makeTempDir({
    prefix: "deno_couscous_test_",
  });
  await cb(tempDir);
  await Deno.remove(tempDir, { recursive: true });
};

export const withSource = (
  cb: (tempSrc: SourceFile) => void | Promise<void>,
  manipulationSettings: Partial<ManipulationSettings> = {
    indentationText: IndentationText.TwoSpaces,
  }
) =>
  withDir(async (tempDir) => {
    const tempFile = await Deno.makeTempFile({ dir: tempDir });
    const project = new Project({
      libFolderPath: tempDir,
      manipulationSettings,
    });
    const source = project.addSourceFileAtPath(tempFile);
    await cb(source);
  });

type TestParameter = Deno.TestContext | TestCallback;
type TestCallback = (
  tempSrc: SourceFile,
  context: Deno.TestContext
) => void | Promise<void>;

export function test(name: string, cb: TestCallback): void;
export function test(
  name: string,
  context: Deno.TestContext,
  cb: TestCallback
): Promise<boolean>;
export function test(name: string, ...args: TestParameter[]) {
  if (typeof args[0] === "function") {
    const cb = args[0] as TestCallback;
    Deno.test(name, { permissions: { read: true, write: true } }, (t) =>
      withSource((source) => cb(source, t))
    );
  } else {
    const context = args[0] as Deno.TestContext;
    const cb = args[1] as TestCallback;
    return context.step(name, (t) => withSource((source) => cb(source, t)));
  }
}
