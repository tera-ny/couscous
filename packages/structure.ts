import {
  InterfaceDeclarationStructure,
  FunctionDeclarationStructure,
  TypeAliasDeclarationStructure,
  StructureKind,
} from "https://deno.land/x/ts_morph@14.0.0/mod.ts";
import { Route, ParameterType } from "./type/index.ts";

export const RouteOptionStructure: InterfaceDeclarationStructure = {
  kind: StructureKind.Interface,
  name: "RouteOption",
  properties: [
    {
      name: "query",
      type: "{ [key: string]: string }",
      hasQuestionToken: true,
    },
    {
      name: "fragment",
      type: "string",
      hasQuestionToken: true,
    },
  ],
};

export const ParameterTypeStructure: TypeAliasDeclarationStructure = {
  kind: StructureKind.TypeAlias,
  name: "ParameterType",
  type: [
    ParameterType.Single,
    ParameterType.Rest,
    ParameterType.OptionalRest,
  ].join(" | "),
};

export const RouteFunctionStructure: FunctionDeclarationStructure = {
  kind: StructureKind.Function,
  name: "route",
  isExported: true,
  parameters: [
    { name: "identity", type: "Identity" },
    {
      name: "...args",
      type: "ParameterType[]",
    },
  ],
};

export const IdentityTypeStructure = (
  routes: [Pick<Route, "identity">, ...Pick<Route, "identity">[]]
): TypeAliasDeclarationStructure => {
  if (!routes.length) throw "routes is empty";
  return {
    kind: StructureKind.TypeAlias,
    name: "Identity",
    type: (writer) => {
      routes.forEach((route, index) => {
        writer.write(`"${route.identity}"`);
        if (routes.length > index + 1) writer.write(" | ");
      });
    },
  };
};
