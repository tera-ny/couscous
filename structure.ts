import {
  InterfaceDeclarationStructure,
  FunctionDeclarationStructure,
  StructureKind,
} from "ts_morph";

export const RouteOptionStructure: InterfaceDeclarationStructure = {
  kind: StructureKind.Interface,
  name: "RouteOption",
  properties: [
    {
      name: "query",
      type: "{ [key: string]: string }",
      hasQuestionToken: true,
    },
  ],
};

export const RouteFunctionStructure: FunctionDeclarationStructure = {
  kind: StructureKind.Function,
  name: "route",
  isExported: true,
  parameters: [
    { name: "identity", type: "string" },
    { name: "option", type: "RouteOption", hasQuestionToken: true },
  ],
};
