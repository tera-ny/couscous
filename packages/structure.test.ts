import { test } from "./_test_helper/fileSystem.ts";
import { RouteOptionStructure, IdentityTypeStructure } from "./structure.ts";
import { assertSnapshot } from "./_test_helper/equal.ts";

test("RouteOption with snapshot", async (source) => {
  source.addInterface(RouteOptionStructure);
  await source.save();
  await assertSnapshot(
    source.getFilePath(),
    "_snapshots/structure_test_routeoption.snapshot"
  );
});

test("IdentityType with snapshot", async (source) => {
  source.addTypeAlias(
    IdentityTypeStructure([
      { identity: "foo" },
      { identity: "bar" },
      { identity: "hoge" },
      { identity: "piyo" },
      { identity: "fuga/${id}" },
    ])
  );
  await source.save();
  await assertSnapshot(
    source.getFilePath(),
    "_snapshots/structure_test_identitytype.snapshot"
  );
});
