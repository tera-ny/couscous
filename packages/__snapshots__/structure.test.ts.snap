export const snapshot = {};

snapshot[`RouteOption with snapshot 1`] = `"interface RouteOption {\\n  query?: { [key: string]: string };\\n  fragment?: string;\\n}\\n"`;

snapshot[`ParameterType with snapshot 1`] = `"type ParameterType = string | [string, ...string[]] | string[] | RouteOption | undefined;\\n"`;

snapshot[`IdentityType with snapshot 1`] = `'type Identity = "foo" | "bar" | "hoge" | "piyo" | "fuga/\${id}";\\n'`;
